drapergem/draper
# Add support for Rails 5 API-only applications

## Description
 Draper expects your `ApplicationController` to include `ActionView::Rendering`. The
`ApplicationController` generated by Rails 5 API-only applications (created with
`rails new --api`) doesn't by default. However, including `ActionView::Rendering` in
 `ApplicatonController` breaks `render :json` due to `render_to_body` being overridden.

This compatibility patch fixes the issue by restoring the original `render_to_body`
method after including `ActionView::Rendering`. 

Originally, I was going to provide documentation in the README on how to apply this patch. However, I ended up deciding to build it in. That way Draper works out of the box for api only applications at the expense of including `ActionView::Rendering` in `ApplicationController`.

Ultimately, including `ActionView::Rendering`in an `ActionController::API` may not be supported functionality by Rails (see Rails issue for more detail: https://github.com/rails/rails/issues/27211). This hack is meant to be a temporary solution until we can find a way to not rely on the controller layer.

## Testing

1. Create a new Rails 5 app: `rails new api_only --api`
1. Add Draper to the Gemfile: `gem 'draper', github: 'drapergem/draper', branch: 'feature/rails-5-api-only-support'`
1. After bundling, run `rails c`. This should no longer throw an exception.
1. Generate scaffold with decorator: `rails g scaffold Post title description`
1. Verify you can (1) decorate the object and (2) `render :json`.

## To-Dos

## References
* [Draper rails-5 + ActionController::API (Draper Issue #753)](https://github.com/drapergem/draper/issues/753)
* [Support Rails 5 API-only apps (Draper Pull Request #766)](https://github.com/drapergem/draper/pull/766)
* [Including ActionView::Rendering in API controller without jbuilder breaks render :json (Rails Issue #27211)](https://github.com/rails/rails/issues/27211)
